### ISSUE TITLE
training coco2017 from pretrained model get nan loss ?

### ISSUE THREAD

**Issue by XiaoLaoDi** [zjhuang22/maskscoring_rcnn#53]
I am training coco2017, and changed the batchsize from default 16 to 4, but I got an nan loss after a few hundred normal iterations ? So what's wrong with that ?

**Comment by XiaoLaoDi** [comment]
Here is the config file: e2e_ms_rcnn_R_50_FPN_1x.yaml
MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: "catalog://ImageNetPretrained/MSRA/R-50"
  PRETRAINED_MODELS: 'pretrained_models'
  BACKBONE:
    CONV_BODY: "R-50-FPN"
    OUT_CHANNELS: 256
  RPN:
    USE_FPN: True
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    PRE_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TEST: 1000
  ROI_HEADS:
    USE_FPN: True
  ROI_BOX_HEAD:
    POOLER_RESOLUTION: 7
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POOLER_SAMPLING_RATIO: 2
    FEATURE_EXTRACTOR: "FPN2MLPFeatureExtractor"
    PREDICTOR: "FPNPredictor"
  ROI_MASK_HEAD:
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    FEATURE_EXTRACTOR: "MaskRCNNFPNFeatureExtractor"
    PREDICTOR: "MaskRCNNC4Predictor"
    POOLER_RESOLUTION: 14
    POOLER_SAMPLING_RATIO: 2
    RESOLUTION: 28
    SHARE_BOX_FEATURE_EXTRACTOR: False
  MASK_ON: True
  MASKIOU_ON: True
DATASETS:
  TRAIN: ("coco_2017_train",)
  TEST: ("coco_2017_val",)
DATALOADER:
  SIZE_DIVISIBILITY: 32
SOLVER:
  IMS_PER_BATCH: 4
  BASE_LR: 0.02
  WEIGHT_DECAY: 0.0001
  STEPS: (60000, 80000)
  MAX_ITER: 90000
TEST:
  EXPECTED_RESULTS: []
  EXPECTED_RESULTS_SIGMA_TOL: 4
  IMS_PER_BATCH: 1

**Comment by XiaoLaoDi** [comment]
some of the training log like this: 
2019-06-13 12:40:33,705 maskrcnn_benchmark.trainer INFO: eta: 19:16:58  iter: 3700  loss: 1.3417 (1.3007)  loss_classifier: 0.3692 (0.4179)  loss_box_reg: 0.1055 (0.1274)  loss_mask: 0.4583 (0.4700)  loss_maskiou: 0.0126 (0.0134)  loss_objectness: 0.2127 (0.1627)  loss_rpn_box_reg: 0.1027 (0.1093)  time: 0.7727 (0.8044)  data: 0.0216 (0.0237)  lr: 0.020000  max mem: 6019
2019-06-13 12:40:49,454 maskrcnn_benchmark.trainer INFO: eta: 19:16:34  iter: 3720  loss: 1.0314 (1.2994)  loss_classifier: 0.3077 (0.4175)  loss_box_reg: 0.1015 (0.1272)  loss_mask: 0.3828 (0.4696)  loss_maskiou: 0.0084 (0.0134)  loss_objectness: 0.1371 (0.1626)  loss_rpn_box_reg: 0.0725 (0.1092)  time: 0.7803 (0.8043)  data: 0.0172 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:41:05,364 maskrcnn_benchmark.trainer INFO: eta: 19:16:13  iter: 3740  loss: 1.0711 (1.2987)  loss_classifier: 0.3263 (0.4171)  loss_box_reg: 0.0873 (0.1271)  loss_mask: 0.4198 (0.4693)  loss_maskiou: 0.0107 (0.0134)  loss_objectness: 0.1524 (0.1625)  loss_rpn_box_reg: 0.0725 (0.1092)  time: 0.7858 (0.8042)  data: 0.0166 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:41:21,689 maskrcnn_benchmark.trainer INFO: eta: 19:16:03  iter: 3760  loss: 1.0745 (1.2985)  loss_classifier: 0.3665 (0.4170)  loss_box_reg: 0.1128 (0.1271)  loss_mask: 0.4073 (0.4690)  loss_maskiou: 0.0116 (0.0134)  loss_objectness: 0.1331 (0.1628)  loss_rpn_box_reg: 0.0779 (0.1092)  time: 0.8190 (0.8043)  data: 0.0213 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:41:36,840 maskrcnn_benchmark.trainer INFO: eta: 19:15:25  iter: 3780  loss: 1.1817 (1.2985)  loss_classifier: 0.3834 (0.4168)  loss_box_reg: 0.0837 (0.1269)  loss_mask: 0.4441 (0.4690)  loss_maskiou: 0.0125 (0.0134)  loss_objectness: 0.1632 (0.1630)  loss_rpn_box_reg: 0.1057 (0.1094)  time: 0.7446 (0.8041)  data: 0.0215 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:41:53,078 maskrcnn_benchmark.trainer INFO: eta: 19:15:13  iter: 3800  loss: 1.5016 (1.2995)  loss_classifier: 0.4001 (0.4171)  loss_box_reg: 0.1276 (0.1270)  loss_mask: 0.4638 (0.4689)  loss_maskiou: 0.0109 (0.0133)  loss_objectness: 0.1688 (0.1633)  loss_rpn_box_reg: 0.1101 (0.1098)  time: 0.8080 (0.8041)  data: 0.0216 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:42:05,988 maskrcnn_benchmark.trainer INFO: eta: 19:13:45  iter: 3820  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.6314 (0.8033)  data: 0.0210 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:42:18,469 maskrcnn_benchmark.trainer INFO: eta: 19:12:09  iter: 3840  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.6150 (0.8023)  data: 0.0215 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:42:30,752 maskrcnn_benchmark.trainer INFO: eta: 19:10:29  iter: 3860  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.6018 (0.8014)  data: 0.0181 (0.0236)  lr: 0.020000  max mem: 6019
2019-06-13 12:42:43,080 maskrcnn_benchmark.trainer INFO: eta: 19:08:51  iter: 3880  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.6107 (0.8004)  data: 0.0203 (0.0236)  lr: 0.020000  max mem: 6019

**Comment by XiaoLaoDi** [comment]
when I change the model to : configs/e2e_ms_rcnn_R_101_FPN_1x.yaml
I got the following nan loss log:
2019-06-13 13:25:39,585 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:39:01  iter: 620  loss: 1.3794 (1.4556)  loss_classifier: 0.4094 (0.3740)  loss_box_reg: 0.1194 (0.0792)  loss_mask: 0.5691 (0.6480)  loss_maskiou: 0.0119 (0.0222)  loss_objectness: 0.1251 (0.2152)  loss_rpn_box_reg: 0.0479 (0.1170)  time: 1.0277 (0.9929)  data: 0.0189 (0.0255)  lr: 0.020000  max mem: 7234
2019-06-13 13:25:59,876 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:39:42  iter: 640  loss: 1.5185 (1.4561)  loss_classifier: 0.3666 (0.3740)  loss_box_reg: 0.0803 (0.0797)  loss_mask: 0.5650 (0.6460)  loss_maskiou: 0.0131 (0.0220)  loss_objectness: 0.2094 (0.2152)  loss_rpn_box_reg: 0.1754 (0.1192)  time: 1.0016 (0.9935)  data: 0.0201 (0.0255)  lr: 0.020000  max mem: 7234
2019-06-13 13:26:20,281 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:40:34  iter: 660  loss: 1.2783 (1.4564)  loss_classifier: 0.3499 (0.3748)  loss_box_reg: 0.0983 (0.0804)  loss_mask: 0.5796 (0.6446)  loss_maskiou: 0.0124 (0.0218)  loss_objectness: 0.1189 (0.2149)  loss_rpn_box_reg: 0.0398 (0.1199)  time: 1.0016 (0.9943)  data: 0.0189 (0.0254)  lr: 0.020000  max mem: 7314
2019-06-13 13:26:39,320 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:38:23  iter: 680  loss: 5.2807 (nan)  loss_classifier: 0.7110 (nan)  loss_box_reg: 0.1448 (nan)  loss_mask: 0.7507 (nan)  loss_maskiou: 0.0635 (nan)  loss_objectness: 2.0321 (nan)  loss_rpn_box_reg: 1.7623 (nan)  time: 0.9243 (0.9931)  data: 0.0198 (0.0254)  lr: 0.020000  max mem: 7314
2019-06-13 13:26:56,157 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:31:37  iter: 700  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.8401 (0.9888)  data: 0.0221 (0.0253)  lr: 0.020000  max mem: 7314
2019-06-13 13:27:13,447 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:26:09  iter: 720  loss: nan (nan)  loss_classifier: nan (nan)  loss_box_reg: nan (nan)  loss_mask: nan (nan)  loss_maskiou: nan (nan)  loss_objectness: nan (nan)  loss_rpn_box_reg: nan (nan)  time: 0.8604 (0.9853)  data: 0.0228 (0.0252)  lr: 0.020000  max mem: 7314

**Comment by zjhuang22** [comment]
What is your training command? May your batchsize is too small.

**Comment by XiaoLaoDi** [comment]
@zjhuang22 

> What is your training command? May your batchsize is too small.

Yes, maybe the batchsize is too small, when I change it from 4 to 18, the nan error disappear, and all the training is OK. But why the small batchsize leads to the nan loss ?

**Comment by zjhuang22** [comment]
Small batch leads to gradient instability.

**Comment by XiaoLaoDi** [comment]
@zjhuang22 

> Small batch leads to gradient instability.

OK, thanks.
